File: mobile/src/main/java/org/openhab/habdroid/ui/CloudNotificationAdapter.java
/*
 * Copyright (c) 2010-2016, openHAB.org and others.
 *
 *   All rights reserved. This program and the accompanying materials
 *   are made available under the terms of the Eclipse Public License v1.0
 *   which accompanies this distribution, and is available at
 *   http://www.eclipse.org/legal/epl-v10.html
 */

package org.openhab.habdroid.ui;

import android.content.Context;
import android.net.Uri;
import android.text.TextUtils;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.core.graphics.drawable.DrawableCompat;
import androidx.recyclerview.widget.RecyclerView;

import org.openhab.habdroid.R;
import org.openhab.habdroid.core.connection.Connection;
import org.openhab.habdroid.core.connection.ConnectionFactory;
import org.openhab.habdroid.model.CloudNotification;
import org.openhab.habdroid.ui.widget.WidgetImageView;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class CloudNotificationAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    public interface LoadMoreListener {
        void loadMoreItems();
    }

    private static final int VIEW_TYPE_NOTIFICATION = 0;
    private static final int VIEW_TYPE_LOADING = 1;

    private final ArrayList<CloudNotification> mItems = new ArrayList<>();
    private final LayoutInflater mInflater;
    private final LoadMoreListener mLoadMoreListener;
    private boolean mHasMoreItems;
    private boolean mWaitingForMoreData;
    private int mHighlightedPosition = -1;

    public CloudNotificationAdapter(Context context, LoadMoreListener loadMoreListener) {
        super();
        mInflater = LayoutInflater.from(context);
        mLoadMoreListener = loadMoreListener;
        mHasMoreItems = false;
    }

    public void addLoadedItems(List<CloudNotification> items, boolean hasMoreItems) {
        mItems.addAll(items);
        mHasMoreItems = hasMoreItems;
        mWaitingForMoreData = false;
        notifyDataSetChanged();
    }

    public void clear() {
        mItems.clear();
        mHasMoreItems = false;
        mWaitingForMoreData = false;
        notifyDataSetChanged();
    }

    public int findPositionForId(String id) {
        for (int i = 0; i < mItems.size(); i++) {
            if (TextUtils.equals(mItems.get(i).id(), id)) {
                return i;
            }
        }
        return -1;
    }

    public void highlightItem(int position) {
        mHighlightedPosition = position;
        notifyItemChanged(position);
    }

    @Override
    public int getItemCount() {
        return mItems.size() + (mHasMoreItems ? 1 : 0);
    }

    @Override
    public int getItemViewType(int position) {
        return position == mItems.size() ? VIEW_TYPE_LOADING : VIEW_TYPE_NOTIFICATION;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_LOADING) {
            return new LoadingIndicatorViewHolder(mInflater, parent);
        } else {
            return new NotificationViewHolder(mInflater, parent);
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof NotificationViewHolder) {
            NotificationViewHolder nvh = (NotificationViewHolder) holder;
            nvh.bind(mItems.get(position));
        } else {
            // loading indicator
            holder.itemView.setVisibility(mHasMoreItems ? View.VISIBLE : View.GONE);
            if (mHasMoreItems && !mWaitingForMoreData) {
                mLoadMoreListener.loadMoreItems();
                mWaitingForMoreData = true;
            }
        }

        if (position == mHighlightedPosition) {
            final View v = holder.itemView;
            v.post(() -> {
                if (v.getBackground() != null) {
                    final int centerX = v.getWidth() / 2;
                    final int centerY = v.getHeight() / 2;
                    DrawableCompat.setHotspot(v.getBackground(), centerX, centerY);
                }
                v.setPressed(true);
                v.setPressed(false);
                mHighlightedPosition = -1;
            });
        }
    }

    public static class NotificationViewHolder extends RecyclerView.ViewHolder {
        final TextView mCreatedView;
        final TextView mMessageView;
        final WidgetImageView mIconView;
        final TextView mSeverityView;

        public NotificationViewHolder(LayoutInflater inflater, ViewGroup parent) {
            super(inflater.inflate(R.layout.notificationlist_item, parent, false));
            mCreatedView = itemView.findViewById(R.id.notificationCreated);
            mMessageView = itemView.findViewById(R.id.notificationMessage);
            mSeverityView = itemView.findViewById(R.id.notificationSeverity);
            mIconView = itemView.findViewById(R.id.notificationImage);
        }

        public void bind(CloudNotification notification) {
            mCreatedView.setText(DateUtils.getRelativeDateTimeString(mCreatedView.getContext(),
                    notification.createdTimestamp(),
                    DateUtils.MINUTE_IN_MILLIS, DateUtils.WEEK_IN_MILLIS, 0));
            mMessageView.setText(notification.message());

            if (notification.icon() != null) {
                Connection conn = ConnectionFactory.getConnection(Connection.TYPE_CLOUD);
                String iconUrl = String.format(Locale.US, "images/%s.png",
                        Uri.encode(notification.icon()));
                mIconView.setImageUrl(conn, iconUrl, 2000);
            } else {
                mIconView.setImageResource(R.drawable.ic_openhab_appicon_24dp);
            }
            mSeverityView.setText(notification.severity());
            mSeverityView.setVisibility(
                    TextUtils.isEmpty(notification.severity()) ? View.GONE : View.VISIBLE);
        }
    }

    public static class LoadingIndicatorViewHolder extends RecyclerView.ViewHolder {
        public LoadingIndicatorViewHolder(LayoutInflater inflater, ViewGroup parent) {
            super(inflater.inflate(R.layout.notificationlist_loading_item, parent, false));
        }
    }
}

File: mobile/src/main/java/org/openhab/habdroid/ui/CloudNotificationAdapter.kt
/*
 * Copyright (c) 2010-2024 Contributors to the openHAB project
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package org.openhab.habdroid.ui

import android.content.Context
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.view.isGone
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.openhab.habdroid.R
import org.openhab.habdroid.core.connection.ConnectionFactory
import org.openhab.habdroid.model.CloudMessage
import org.openhab.habdroid.ui.widget.WidgetImageView
import org.openhab.habdroid.util.determineDataUsagePolicy

class CloudNotificationAdapter(context: Context, private val loadMoreListener: () -> Unit) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private val items = mutableListOf<CloudMessage.CloudNotification>()
    private val existingReferenceIds = mutableSetOf<String>()
    private val inflater = LayoutInflater.from(context)
    private var hasMoreItems: Boolean = false
    private var waitingForMoreData: Boolean = false
    private var highlightedPosition = -1

    fun addLoadedItems(loaded: List<CloudMessage.CloudNotification>, hasMoreItems: Boolean) {
        val existingItemCount = items.size
        val relevant = loaded.filter {
            // Collapse multiple notifications with the same reference ID into the latest one by accepting either
            // - notifications without reference ID or
            // - notifications whose reference ID we haven't seen yet
            it.id.referenceId == null || existingReferenceIds.add(it.id.referenceId)
        }
        items.addAll(relevant)
        notifyItemRangeInserted(existingItemCount, relevant.size)
        if (this.hasMoreItems && !hasMoreItems) {
            notifyItemRemoved(items.size)
        } else if (!this.hasMoreItems && hasMoreItems) {
            notifyItemInserted(items.size)
        }
        this.hasMoreItems = hasMoreItems
        waitingForMoreData = false
    }

    fun clear() {
        val existingItemCount = itemCount
        items.clear()
        existingReferenceIds.clear()
        hasMoreItems = false
        waitingForMoreData = false
        notifyItemRangeRemoved(0, existingItemCount)
    }

    fun findPositionForId(id: String): Int = items.indexOfFirst { item -> item.id.persistedId == id }

    fun highlightItem(position: Int) {
        highlightedPosition = position
        notifyItemChanged(position)
    }

    override fun getItemCount(): Int = items.size + if (hasMoreItems) 1 else 0

    override fun getItemViewType(position: Int): Int =
        if (position == items.size) VIEW_TYPE_LOADING else VIEW_TYPE_NOTIFICATION

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
        if (viewType == VIEW_TYPE_LOADING) {
            LoadingIndicatorViewHolder(inflater, parent)
        } else {
            NotificationViewHolder(inflater, parent)
        }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is NotificationViewHolder) {
            holder.bind(items[position])
        } else {
            // loading indicator
            holder.itemView.isVisible = hasMoreItems
            if (hasMoreItems && !waitingForMoreData) {
                loadMoreListener()
                waitingForMoreData = true
            }
        }

        if (position == highlightedPosition) {
            holder.itemView.playPressAnimationAndCallBack {
                highlightedPosition = -1
            }
        }
    }

    class NotificationViewHolder(inflater: LayoutInflater, parent: ViewGroup) :
        RecyclerView.ViewHolder(inflater.inflate(R.layout.notificationlist_item, parent, false)) {
        private val titleView: TextView = itemView.findViewById(R.id.notificationTitle)
        private val messageView: TextView = itemView.findViewById(R.id.notificationMessage)
        private val createdView: TextView = itemView.findViewById(R.id.notificationCreated)
        private val iconView: WidgetImageView = itemView.findViewById(R.id.notificationIcon)
        private val imageView: WidgetImageView = itemView.findViewById(R.id.notificationImage)
        private val tagView: TextView = itemView.findViewById(R.id.notificationTag)

        fun bind(notification: CloudMessage.CloudNotification) {
            createdView.text = DateUtils.getRelativeDateTimeString(
                itemView.context,
                notification.createdTimestamp,
                DateUtils.MINUTE_IN_MILLIS,
                DateUtils.WEEK_IN_MILLIS,
                0
            )
            titleView.text = notification.title
            titleView.isVisible = notification.title.isNotEmpty()
            messageView.text = notification.message
            messageView.isVisible = notification.message.isNotEmpty()

            val conn = ConnectionFactory.activeCloudConnection?.connection
            if (conn == null) {
                iconView.applyFallbackDrawable()
                imageView.isVisible = false
            } else {
                if (notification.icon != null) {
                    iconView.setImageUrl(
                        conn,
                        notification.icon.toUrl(
                            itemView.context,
                            itemView.context.determineDataUsagePolicy(conn).loadIconsWithState
                        ),
                        timeoutMillis = 2000
                    )
                } else {
                    iconView.applyFallbackDrawable()
                }
                imageView.isVisible = notification.mediaAttachmentUrl != null
                CoroutineScope(Dispatchers.IO + Job()).launch {
                    val bitmap = notification.loadImage(conn, itemView.context, itemView.width)
                    withContext(Dispatchers.Main) {
                        imageView.setImageBitmap(bitmap)
                    }
                }
            }
            tagView.text = notification.tag
            tagView.isGone = notification.tag.isNullOrEmpty()
        }
    }

    class LoadingIndicatorViewHolder(inflater: LayoutInflater, parent: ViewGroup) :
        RecyclerView.ViewHolder(inflater.inflate(R.layout.notificationlist_loading_item, parent, false))

    companion object {
        private const val VIEW_TYPE_NOTIFICATION = 0
        private const val VIEW_TYPE_LOADING = 1
    }
}
